crates/circuit/src/circuit_data.rs

```diff
@@ -744,184 +745,130 @@ impl CircuitData {
-                        self.data.remove(index);
-                        self.reindex_parameter_table(py)?;
-                    }
-                    Ok(())
-                } else {
-                    Err(PyIndexError::new_err(format!(
-                        "No element at index {:?} in circuit data",
-                        index
-                    )))
-                }
-            }
-        }
+    pub fn __delitem__(&mut self, py: Python, index: PySequenceIndex) -> PyResult<()> {
+        self.delitem(py, index.with_len(self.data.len())?)
     }
 
     pub fn setitem_no_param_table_update(
         &mut self,
-        py: Python<'_>,
-        index: isize,
-        value: &Bound<PyAny>,
+        index: usize,
+        value: PyRef<CircuitInstruction>,
     ) -> PyResult<()> {
-        let index = self.convert_py_index(index)?;
-        let value: PyRef<CircuitInstruction> = value.downcast()?.borrow();
-        let mut packed = self.pack(py, value)?;
+        let mut packed = self.pack(value)?;
         std::mem::swap(&mut packed, &mut self.data[index]);
         Ok(())
     }
 
-    pub fn __setitem__(
-        &mut self,
-        py: Python<'_>,
-        index: SliceOrInt,
-        value: &Bound<PyAny>,
-    ) -> PyResult<()> {
-        match index {
-            SliceOrInt::Slice(slice) => {
-                let indices = slice.indices(self.data.len().try_into().unwrap())?;
-                let slice = self.convert_py_slice(&slice)?;
-                let values = value.iter()?.collect::<PyResult<Vec<Bound<PyAny>>>>()?;
-                if indices.step != 1 && slice.len() != values.len() {
-                    // A replacement of a different length when step isn't exactly '1'
-                    // would result in holes.
-                    return Err(PyValueError::new_err(format!(
-                        "attempt to assign sequence of size {:?} to extended slice of size {:?}",
-                        values.len(),
-                        slice.len(),
-                    )));
-                }
+    pub fn __setitem__(&mut self, index: PySequenceIndex, value: &Bound<PyAny>) -> PyResult<()> {
+        fn set_single(slf: &mut CircuitData, index: usize, value: &Bound<PyAny>) -> PyResult<()> {
+            let py = value.py();
+            let mut packed = slf.pack(value.downcast::<CircuitInstruction>()?.borrow())?;
+            slf.remove_from_parameter_table(py, index)?;
+            std::mem::swap(&mut packed, &mut slf.data[index]);
+            slf.update_param_table(py, index, None)?;
+            Ok(())
+        }
 
-                for (i, v) in slice.iter().zip(values.iter()) {
-                    self.__setitem__(py, SliceOrInt::Int(*i), v)?;
+        let py = value.py();
+        match index.with_len(self.data.len())? {
+            SequenceIndex::Int(index) => set_single(self, index, value),
+            indices @ SequenceIndex::PosRange {
```

Comment:
I think this is the first time I've seen the match binding syntax used in practice.

Reply:
I think I used it in the OQ2 lexer as well - I don't use it often, but it's occasionally useful.